contrib/migrate/migrate.c

```diff
@@ -0,0 +1,841 @@
+      error = -1;
+    }
+
+  cub_db_query_end = dlsym (dl_handle, "db_query_end");
+  if (cub_db_query_end == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  cub_db_commit_transaction = dlsym (dl_handle, "db_commit_transaction");
+  if (cub_db_commit_transaction == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  cub_db_abort_transaction = dlsym (dl_handle, "db_abort_transaction");
+  if (cub_db_abort_transaction == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  cub_db_shutdown = dlsym (dl_handle, "db_shutdown");
+  if (cub_db_shutdown == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  cub_db_error_string = dlsym (dl_handle, "db_error_string");
+  if (cub_db_error_string == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  cub_Au_disable = dlsym (dl_handle, "Au_disable");
+  if (cub_Au_disable == NULL)
+    {
+      PRINT_LOG ("%s", dlerror ());
+      error = -1;
+    }
+
+  return error;
+}
+
+static int
+migrate_execute_query (const char *query)
+{
+  DB_QUERY_RESULT *result;
+  int error;
+
+  error = cub_db_execute_query (query, &result);
+  printf ("migrate: execute query: %s\n", query);
+  if (error < 0)
+    {
+      printf ("migrate: execute query failed \"%s\"\n", query);
+      return -1;
+    }
+  cub_db_query_end (result);
+}
+
+static int
+migrate_generated (const char *generated, int col_num)
+{
+  DB_QUERY_RESULT *gen_result;
+  DB_QUERY_RESULT *result;
+  const char *query;
+  int error;
+
+  error = cub_db_execute_query (generated, &gen_result);
+  if (error < 0)
+    {
+      printf ("generated: execute query failed\n \"%s\"\n", generated);
+      return -1;
+    }
+
+  if (cub_db_query_first_tuple (gen_result) == DB_CURSOR_SUCCESS)
+    {
+      do
+	{
+	  int i;
+	  DB_VALUE value;
+
+	  for (i = 0; i < col_num; i++)
+	    {
+	      /* from query */
+	      error = cub_db_query_get_tuple_value (gen_result, i, &value);
+	      if (error < 0)
+		{
+		  printf ("generated: can not get a tuple for \"%s\"\n", generated);
+		  return -1;
+		}
+
+	      query = value.data.ch.medium.buf;
+
+	      /* from generated result */
+	      error = migrate_execute_query (query);
+	      if (error < 0)
+		{
+		  return error;
+		}
+	    }
+	  error = cub_db_query_next_tuple (gen_result);
+	}
+      while (error != DB_CURSOR_END && error != DB_CURSOR_ERROR);
+
+      cub_db_query_end (gen_result);
+    }
+  else
+    {
+      printf ("generated: can not get first tuple for \"%s\"\n", generated);
+      return -1;
+    }
+
+  if (error < 0)
+    {
+      printf ("generated: can not get a next tuple for \"%s\"\n", generated);
+      return -1;
+    }
+
+  return 0;
+}
+
+static int
+migrate_queries ()
+{
+  DB_QUERY_RESULT *result;
+  DB_VALUE value;
+  int i, error;
+
+  /* catalog query */
+  for (i = 0; i < sizeof (catalog_query) / sizeof (const char *); i++)
```

Comment:
"catalog_query" must be executed after "rename_query".

Reply:
Thanks, fixed.